nbs/01_io.ipynb

```diff
@@ -266,6 +268,151 @@
+    "            raise NotImplementedError(\"Maxdiff not implemented yet\")\n",
+    "        else:\n",
+    "            raise NotImplementedError(f\"Create block {dict_with_create['type']} not supported\")\n",
+    "    return zip(dfs_new, metas_new)\n",
+    "\n",
+    "\n",
+    "def has_create(structure):\n",
+    "    return any(filter(lambda s: 'create' in s.keys(), structure))\n",
+    "\n",
+    "\n",
+    "def add_topk_structure(df, dict_with_create):\n",
+    "    \"\"\"\n",
+    "    agg_index = -1  index of the from_cols group to be aggregated.\n",
+    "                    Note that 1 is first group, -1 to last, 0 not allowed.\n",
+    "    \"\"\"\n",
+    "    create = dict_with_create['create']\n",
+    "    name = create['name']\n",
+    "    scale = create.get('scale', {})\n",
+    "    r = re.compile(create['from_columns'])\n",
+    "    agg_ind = create.get('agg_index', -1)\n",
+    "    agg_ind = agg_ind-1 if agg_ind > 0 else agg_ind\n",
+    "    from_cols = list(filter(lambda s: r.match(s), df.columns))\n",
+    "    no_groups = len(r.match(from_cols[0]).groups())\n",
+    "    has_subgroups = no_groups >= 2\n",
+    "    template = create.get('res_cols', '')\n",
+    "    kmax = int(create['k']) if create.get('k','max') != 'max' else 'max'\n",
+    "    na_vals = create.get('na_vals', [])\n",
+    "    if has_subgroups:\n",
+    "        def get_groups(column):\n",
+    "            gs = list(r.match(column).groups())\n",
+    "            gs.pop(agg_ind)\n",
+    "            return gs\n",
+    "        subgroups_ids = np.unique(np.vstack(list(map(get_groups,from_cols))),axis=0)\n",
+    "        subgroups = [[col for col in from_cols if get_groups(col)==g] for g in subgroups_ids]\n",
+    "    else: subgroups = [from_cols]\n",
+    "    topk_dfs, subgroup_metas = [], []\n",
+    "    for subgroup in subgroups:\n",
+    "        subgroup_df = create_topk_dataframe(\n",
+    "            df[subgroup].copy(deep=True), template, kmax, na_vals, agg_ind, r\n",
+    "        )\n",
+    "        topk_dfs.append(subgroup_df)\n",
+    "        if has_subgroups:\n",
+    "            subgroup_name = '_'+'_'.join(map(str,get_groups(subgroup[0])))\n",
+    "        else: subgroup_name = ''\n",
+    "        meta_subgroup = {\n",
+    "            'name': name+subgroup_name,\n",
+    "            'scale': scale,\n",
+    "            'columns': subgroup_df.columns.tolist()\n",
+    "            }\n",
+    "        if 'name' in subgroups:\n",
+    "            print(f\"Problem in {meta_subgroup}\")\n",
+    "            raise Exception(\"Problem in meta2\")\n",
+    "        subgroup_metas.append(meta_subgroup)\n",
+    "    # created_model = {\"name\": \"imp_top3\", \"structure\": [[[\"A9_R1\", \"A9_R2\", \"A9_R3\"], None]], \"ordered\": False}\n",
+    "    return subgroup_metas, topk_dfs #note: each df has one meta for zip later\n",
+    "\n",
+    "\n",
+    "def create_topk_dataframe(df, template, kmax, na_vals, agg_ind, r):\n",
```

Comment:
I would appreciate a bit more commenting. 

For functions, what is their general goal. 

Inside functions, each logical step could use a comment.
Example "Finding max number of chosen options" for the maxk block etc. 

Without those, your code is very hard to chew through